simpleclient/src/main/java/io/prometheus/client/Histogram.java

```diff
@@ -0,0 +1,262 @@
+        double[] tmp = new double[buckets.length + 1];
+        for (int i = 0; i < buckets.length; i++) {
+          tmp[i] = buckets[i];
+        }
+        tmp[buckets.length] = Double.POSITIVE_INFINITY;
+        buckets = tmp;
+      }
+      dontInitializeNoLabelsChild = true;
+      return new Histogram(this);
+    }
+   
+    /**
+      * Set the upper bounds of buckets for the histogram.
+      */
+    public Builder buckets(double... buckets) {
+      this.buckets = buckets;
+      return this;
+    }
+
+    /**
+      * Set the upper bounds of buckets for the histogram with a linear sequence.
+      */
+    public Builder linearBuckets(double start, double width, int count) {
+      buckets = new double[count];
+      for (int i = 0; i < count; i++){
+        buckets[i] = start + i*width;
+      }
+      return this;
+    }
+    /**
+      * Set the upper bounds of buckets for the histogram with an exponential sequence.
+      */
+    public Builder exponentialBuckets(double start, double factor, int count) {
+      buckets = new double[count];
+      for (int i = 0; i < count; i++) {
+        buckets[i] = start * Math.pow(factor, i);
+      }
+      return this;
+    }
+    
+  }
+
+  /**
+   *  Return a Builder to allow configuration of a new Histogram.
+   */
+  public static Builder build() {
+    return new Builder();
+  }
+
+  @Override
+  protected Child newChild() {
+    return new Child(buckets);
+  }
+
+  /**
+   * Represents an event being timed.
+   */
+  public static class Timer {
+    Child child;
+    long start;
+    private Timer(Child child) {
+      this.child = child;
+      start = Child.timeProvider.nanoTime();
+    }
+    /**
+     * Observe the amount of time in seconds since {@link Child#startTimer} was called.
+     */
+    public void observeDuration() {
+      child.observe((Child.timeProvider.nanoTime() - start) / NANOSECONDS_PER_SECOND);
+    }
+  }
+
+  /**
+   * The value of a single Histogram.
+   * <p>
+   * <em>Warning:</em> References to a Child become invalid after using
+   * {@link SimpleCollector#remove} or {@link SimpleCollector#clear}.
+   */
+  public static class Child {
+    public static class Value {
+      private double sum;  
+      private double[] buckets;
+    }
+
+    private Child(double[] buckets) {
+      upperBounds = buckets;
+      cumulativeCounts = new DoubleAdder[buckets.length];
+      for (int i = 0; i < buckets.length; ++i) {
+        cumulativeCounts[i] = new DoubleAdder();
+      }
+    }
+    private double[] upperBounds;
+    private DoubleAdder[] cumulativeCounts;
+    private DoubleAdder sum = new DoubleAdder();
+
+    static TimeProvider timeProvider = new TimeProvider();
+    /**
+     * Observe the given amount.
+     */
+    public void observe(double amt) {
+      for (int i = 0; i < upperBounds.length; ++i) {
+        // The last bucket is +Inf, so we always increment.
+        if (amt <= upperBounds[i]) {
```

Comment:
The code looks right for the design doc but i do have a question on the design and the fact that bucket bounds must be defined up front. All 3 bucket definition options seem to required this (some both high and low and others just low)..

That's not an easy task and would require trial and error and ongoing tuning to ensure it was a decent histogram?